shell/platform/android/platform_view_android_jni_impl.h

```diff
@@ -82,9 +84,17 @@ class PlatformViewAndroidJNIImpl final : public PlatformViewAndroidJNI {
 
   bool RequestDartDeferredLibrary(int loading_unit_id) override;
 
+  void ClearDirectByteBufferDecodingPreference(
+      const std::string& channel) override;
+
+  void SetDirectByteBufferDecodingPreference(
+      const std::string& channel,
+      bool wants_direct_byte_buffer_for_decoding) override;
+
  private:
   // Reference to FlutterJNI object.
   const fml::jni::JavaObjectWeakGlobalRef java_object_;
+  std::set<std::string> channels_with_indirect_byte_buffer_decoding_;
```

Comment:
To reduce the complexity of the Java/native interface I'd recommend moving the decision of whether to make a copy of the message buffer over to the Java side.

Can `wantsDirectByteBufferForDecoding` be made a method in the `BinaryMessageHandler` interface?

The ability to handle these buffers safely is a property of the `BinaryMessageHandler` implementation, not the channel.

This is an API change, but if most applications are installing message handlers through the channel APIs then this will be transparent.

If `BinaryMessageHandler` API can not be changed, then can the set of channels be moved over to `FlutterJNI`?

Reply:
> To reduce the complexity of the Java/native interface I'd recommend moving the decision of whether to make a copy of the message buffer over to the Java side.

The point where the decision has to be made is in C++ code, it can't live fully in Java.

> The ability to handle these buffers safely is a property of the BinaryMessageHandler implementation, not the channel.

The choice isn't in the channel, it's in the MessageCodec.  The point where the ByteBuffer surfaces from the users perspective is in `MessageCodec.decodeMessage`, that's why the choice exists there.  You want the same person implementing decodeMessage to decide if decodeMessage actually wants a direct ByteBuffer.

Reply:
The C++ side is always allocating a direct buffer and then making a JNI call to either `handleIndirectPlatformMessage` (which makes a copy) or `handlePlatformMessage` (which does not).

Instead, there could be one Java method that checks whether the message handler supports the direct buffers.  If not, it makes a copy before calling the handler.

Reply:
What would the benefit of that be though?

Reply:
The source of the message is from Dart, which goes to C++, the logic has to be in C++.  I think I'm missing something.

Reply:
We talked offline.  I understand now.  The actual preference can be stored in Java, the boolean value on whether to copy or not.  Right now it is needlessly getting sent to C++.  Good idea, making that change.

Reply:
Done, this should make testing the code easier, too.  I'll add the appropriate DartMessenger unit tests tomorrow.